internal/provider/models/sources/webhook.go

```diff
@@ -0,0 +1,91 @@
+package sources
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	. "github.com/hashicorp/terraform-plugin-framework/types"
+	. "github.com/mezmo/terraform-provider-mezmo/internal/client"
+)
+
+type WebhookSourceModel struct {
+	Id              String `tfsdk:"id"`
+	PipelineId      String `tfsdk:"pipeline_id"`
+	Title           String `tfsdk:"title"`
+	Description     String `tfsdk:"description"`
+	GenerationId    Int64  `tfsdk:"generation_id"`
+	GatewayRouteId  String `tfsdk:"gateway_route_id"`
+	SigningKey      String `tfsdk:"signing_key" user_config:"true"`
+	CaptureMetadata Bool   `tfsdk:"capture_metadata" user_config:"true"`
+}
+
+var WebhookSourceResourceSchema = schema.Schema{
+	Description: "Receive data from incoming webhooks using the WebSub protocol",
+	Attributes: ExtendBaseAttributes(map[string]schema.Attribute{
+		"signing_key": schema.StringAttribute{
+			Required:    true,
+			Sensitive:   true,
+			Description: "The key used to sign the body of the request",
+			Validators: []validator.String{
+				stringvalidator.LengthAtLeast(1),
+				stringvalidator.LengthAtMost(512),
+			},
+		},
+	}, []string{"capture_metadata", "gateway_route_id"}),
+}
+
+func WebhookSourceFromModel(plan *WebhookSourceModel, previousState *WebhookSourceModel) (*Source, diag.Diagnostics) {
+	dd := diag.Diagnostics{}
+
+	component := Source{
+		BaseNode: BaseNode{
+			Type:        "webhook",
+			Title:       plan.Title.ValueString(),
+			Description: plan.Description.ValueString(),
+			UserConfig: map[string]any{
+				"signing_key":      plan.SigningKey.ValueString(),
+				"capture_metadata": plan.CaptureMetadata.ValueBool(),
+			},
+		},
+	}
+
+	if previousState == nil {
+		if !plan.GatewayRouteId.IsUnknown() {
+			// Let them specify gateway route id on POST only
+			component.GatewayRouteId = plan.GatewayRouteId.ValueString()
+		}
+	} else {
+		// Set generated fields
+		component.Id = previousState.Id.ValueString()
+		component.GenerationId = previousState.GenerationId.ValueInt64()
+
+		// If they have specified gateway_route_id, then it *cannot* be a different value that what's in state
+		if !plan.GatewayRouteId.IsUnknown() && plan.GatewayRouteId.ValueString() != previousState.GatewayRouteId.ValueString() {
+			details := fmt.Sprintf(
+				"Cannot update \"gateway_route_id\" to %s. This field is immutable after resource creation.",
+				plan.GatewayRouteId,
+			)
+			dd.AddError("Error in plan", details)
+			return nil, dd
+		}
+	}
```

Comment:
Just a note: I see this code in a lot of places. Can we abstract it away into a util of some sort and call it?

Reply:
Yeah, this is related to things with gateway ids.  Now that we have most of them, we can probably extract this out in another card.  I'll investigate.